@feihugis Thanks for these improvements! Great to see so many lines reduced.

The build is failing because of `MapAndBatchDatasetParams`:
`
tensorflow/core/kernels/data/experimental/map_and_batch_dataset_op_test.cc:49:10: error: 'tensorflow::Status tensorflow::data::experimental::{anonymous}::MapAndBatchDatasetParams::GetInputs(const std::vector<tensorflow::Tensor*>&, std::vector<std::unique_ptr<tensorflow::Tensor> >*, absl::InlinedVector<tensorflow::TensorValue, 4>*) const' marked 'override', but does not override
   Status GetInputs(const std::vector<Tensor*>& input_datasets,
`

--- a/tensorflow/core/kernels/data/dataset_test_base.cc
+++ b/tensorflow/core/kernels/data/dataset_test_base.cc
@@ -837,21 +837,28 @@ Status DatasetOpsTestBaseV2::MakeDatasetTensor(
   AttributeVector attributes;
   TF_RETURN_IF_ERROR(dataset_params.GetAttributes(&attributes));
 
-  std::vector<Tensor> inputs;
+  gtl::InlinedVector<TensorValue, 4> inputs;
   for (auto input_dataset : input_datasets) {
-    inputs.emplace_back(*input_dataset);
+    inputs.emplace_back(TensorValue(input_dataset));
   }
   auto input_tensors = dataset_params.GetInputTensors();
-  inputs.insert(inputs.end(), input_tensors.begin(), input_tensors.end());
+  for (auto& input_tensor : input_tensors) {
+    inputs.emplace_back(TensorValue(&input_tensor));
+  }
 
-  GraphConstructorOptions graph_opts;
-  graph_opts.allow_internal_ops = true;
-  graph_opts.expect_device_spec = false;
-  FunctionDef make_dataset_tensor_fdef;
-  TF_RETURN_IF_ERROR(dataset_params.CreateFactory(&make_dataset_tensor_fdef));
-  *dataset = std::make_unique<Tensor>();
-  TF_RETURN_IF_ERROR(RunFunction(make_dataset_tensor_fdef, attributes, inputs,
-                                 graph_opts, /*rets=*/{dataset->get()}));
+  DatasetBase* dataset_base;
+  std::unique_ptr<OpKernel> dataset_kernel;
+  std::unique_ptr<OpKernelContext::Params> dataset_ctx_params;
+  std::unique_ptr<OpKernelContext> dataset_ctx;
+  TF_RETURN_IF_ERROR(MakeDatasetOpKernel(dataset_params, &dataset_kernel));
+  TF_RETURN_IF_ERROR(CreateDatasetContext(dataset_kernel.get(), &inputs,
+                                          &dataset_ctx_params, &dataset_ctx));
+  TF_RETURN_IF_ERROR(
+      CreateDataset(dataset_kernel.get(), dataset_ctx.get(), &dataset_base));
+  Tensor dataset_tensor(DT_VARIANT, TensorShape({}));
+  TF_RETURN_IF_ERROR(
+      StoreDatasetInVariantTensor(dataset_base, &dataset_tensor));
+  *dataset = absl::make_unique<Tensor>(dataset_tensor);
   return Status::OK();
 }
 

--- a/tensorflow/core/kernels/data/dataset_test_base.cc
+++ b/tensorflow/core/kernels/data/dataset_test_base.cc
@@ -773,7 +773,14 @@ Status DatasetOpsTestBaseV2::MakeDataset(
   for (auto input_dataset : input_datasets) {
     inputs.emplace_back(TensorValue(input_dataset));
   }
-  AddTensorInputs(dataset_params.GetInputTensors(), created_tensors, &inputs);
+
+  // Copy the input tensors, storing them in the `inputs` vectors, and storing
+  // owned references to the copies in `created_tensors`.
+  for (auto& input : dataset_params.GetInputTensors()) {
+    auto copy = absl::make_unique<Tensor>(input);
+    inputs.push_back(TensorValue(copy.get()));
+    created_tensors->push_back(std::move(copy));
+  }
 
   TF_RETURN_IF_ERROR(MakeDatasetOpKernel(dataset_params, dataset_kernel));
   TF_RETURN_IF_ERROR(CreateDatasetContext(dataset_kernel->get(), &inputs,
@@ -830,13 +837,12 @@ Status DatasetOpsTestBaseV2::MakeDatasetTensor(
   AttributeVector attributes;
   TF_RETURN_IF_ERROR(dataset_params.GetAttributes(&attributes));
 
-  std::vector<Tensor> dataset_inputs;
+  std::vector<Tensor> inputs;
   for (auto input_dataset : input_datasets) {
-    dataset_inputs.emplace_back(*input_dataset);
+    inputs.emplace_back(*input_dataset);
   }
   auto input_tensors = dataset_params.GetInputTensors();
-  dataset_inputs.insert(dataset_inputs.end(), input_tensors.begin(),
-                        input_tensors.end());
+  inputs.insert(inputs.end(), input_tensors.begin(), input_tensors.end());
 
   GraphConstructorOptions graph_opts;
   graph_opts.allow_internal_ops = true;
@@ -844,21 +850,9 @@ Status DatasetOpsTestBaseV2::MakeDatasetTensor(
   FunctionDef make_dataset_tensor_fdef;
   TF_RETURN_IF_ERROR(dataset_params.CreateFactory(&make_dataset_tensor_fdef));
   *dataset = std::make_unique<Tensor>();
-  TF_RETURN_IF_ERROR(RunFunction(make_dataset_tensor_fdef, attributes,
-                                 dataset_inputs, graph_opts,
-                                 /*rets=*/{dataset->get()}));
+  TF_RETURN_IF_ERROR(RunFunction(make_dataset_tensor_fdef, attributes, inputs,
+                                 graph_opts, /*rets=*/{dataset->get()}));
   return Status::OK();
-}
-
-void DatasetOpsTestBaseV2::AddTensorInputs(
-    const std::vector<Tensor>& tensors,
-    std::vector<std::unique_ptr<Tensor>>* created_tensors,
-    gtl::InlinedVector<TensorValue, 4>* inputs) const {
-  for (auto& input : tensors) {
-    auto copy = absl::make_unique<Tensor>(input);
-    inputs->push_back(TensorValue(copy.get()));
-    created_tensors->push_back(std::move(copy));
-  }
 }
 
 DatasetParams::DatasetParams(DataTypeVector output_dtypes,

--- a/tensorflow/core/kernels/data/dataset_test_base.h
+++ b/tensorflow/core/kernels/data/dataset_test_base.h
@@ -789,12 +789,6 @@ class DatasetOpsTestBaseV2 : public DatasetOpsTestBase {
 
   Status CreateFactory(const DatasetParams& dataset_params,
                        FunctionDef* fdef) const;
-
-  // Copies the given tensors, storing them in the `inputs` vectors, and storing
-  // owned references to the copies in `created_tensors`.
-  void AddTensorInputs(const std::vector<Tensor>& tensors,
-                       std::vector<std::unique_ptr<Tensor>>* created_tensors,
-                       gtl::InlinedVector<TensorValue, 4>* inputs) const;
 };
 
 #define ITERATOR_GET_NEXT_TEST_P(dataset_op_test_class, dataset_params_class, \

--- a/tensorflow/core/kernels/data/experimental/map_and_batch_dataset_op_test.cc
+++ b/tensorflow/core/kernels/data/experimental/map_and_batch_dataset_op_test.cc
@@ -33,10 +33,9 @@ class MapAndBatchDatasetParams : public DatasetParams {
       : DatasetParams(std::move(output_dtypes), std::move(output_shapes),
                       std::move(node_name)),
         other_arguments_(std::move(other_arguments)),
-        batch_size_(CreateTensor<int64>(TensorShape({}), {batch_size})),
-        num_parallel_calls_(
-            CreateTensor<int64>(TensorShape({}), {num_parallel_calls})),
-        drop_remainder_(CreateTensor<bool>(TensorShape({}), {drop_remainder})),
+        batch_size_(batch_size),
+        num_parallel_calls_(num_parallel_calls),
+        drop_remainder_(drop_remainder),
         func_(std::move(func)),
         func_lib_(std::move(func_lib)),
         type_arguments_(std::move(type_arguments)),
@@ -46,14 +45,13 @@ class MapAndBatchDatasetParams : public DatasetParams {
         input_dataset_params.op_name(), input_dataset_params.iterator_prefix());
   }
 
-  Status GetInputs(const std::vector<Tensor*>& input_datasets,
-                   std::vector<std::unique_ptr<Tensor>>* created_tensors,
-                   gtl::InlinedVector<TensorValue, 4>* inputs) const override {
-    inputs->clear();
-    TF_RETURN_IF_ERROR(AddDatasetInputs(input_datasets, 1, inputs));
-    AddTensorInputs({batch_size_, num_parallel_calls_, drop_remainder_},
-                    created_tensors, inputs);
-    return Status::OK();
+  std::vector<Tensor> GetInputTensors() const override {
+    std::vector<Tensor> inputs = other_arguments_;
+    inputs.emplace_back(CreateTensor<int64>(TensorShape({}), {batch_size_}));
+    inputs.emplace_back(
+        CreateTensor<int64>(TensorShape({}), {num_parallel_calls_}));
+    inputs.emplace_back(CreateTensor<bool>(TensorShape({}), {drop_remainder_}));
+    return inputs;
   }
 
   Status GetInputPlaceholder(
@@ -88,9 +86,9 @@ class MapAndBatchDatasetParams : public DatasetParams {
 
  private:
   std::vector<Tensor> other_arguments_;
-  Tensor batch_size_;
-  Tensor num_parallel_calls_;
-  Tensor drop_remainder_;
+  int64 batch_size_;
+  int64 num_parallel_calls_;
+  bool drop_remainder_;
   FunctionDefHelper::AttrValueWrapper func_;
   std::vector<FunctionDef> func_lib_;
   DataTypeVector type_arguments_;

